chore(docs): improve LT and ASOF JOINS documentation (#1406)
questdb/questdb.io: docs/reference/sql/join.md

:::

## ASOF JOIN

> `ASOF JOIN` will join two tables based on their timestamp distance.

Reviewer: ```suggestion
`ASOF JOIN` joins two tables based on their timestamp distance.
```

It can join two different time-series measured at different points in time. For each value
in the first time-series, it takes the most recent value in the second time-series.

> More formally: `ASOF JOIN` will return all records from the left table combined with matching records from the right table. 

Reviewer: What does "more formally" mean here? More precisely?

Author: yes, _more precisely_ is better than _more formally_.  thank you!

This is the idea: ASOF JOIN is not a well-known join type. It's largely specific to time-series databases. We have to educate people about it. And at the same time it's not trivial to explain it. So let's start with something simple and make it more precise gradually. This is the simplest version I could come up with (with @javier's help):
> It can join two different time-series measured at different points in time. For each value
in the first time-series, it takes the most recent value in the second time-series.

This ☝️ is supposed to give you a rough idea of what the ASOF JOIN does. It answer the question _What is it good for?_ This is how I would read it: _"ASOF JOIN is joining timestamps. and it has something to do with "recency"_. 

It does not tell you what exactly is the role of "recency". What are the rules for selecting the record from the right table. Hence the second paragraph:
> More ~formally~ precisely: `ASOF JOIN` returns all records from the left table combined with matching records from the right table. When a record from the left table has multiple matching records in the right table the ASOF JOIN selects only the record with the closest timestamp **prior or equal to** the timestamp of the left record. ASOF JOIN behaves as the LEFT OUTER JOIN with additional time filter so the right table returns at most one record per each record from the left side.

This ☝️ is more precise than the shorter version above. At the same time, it's a mouthful. I wouldn't want to read this as a very first intro to ASOF JOINs. So the short intro tells you: "Why would you want to even consider it". It's basically a motivation. The 2nd paragraph tells you: "How exactly it works?" - giving you the exact rules - when you actually feel ASOIF is what you need.

Does it make any sense to you?:)

> More formally: `ASOF JOIN` will return all records from the left table combined with matching records from the right table. 

Reviewer: ```suggestion
More formally: `ASOF JOIN` returns all records from the left table combined with matching records from the right table. 
```

When a record from the left table has multiple matching records in the right table the ASOF JOIN selects only the record with the closest
timestamp **prior or equal to** the timestamp of the left record. ASOF JOIN behaves as the LEFT OUTER JOIN with additional time filter
so the right table returns at most one record per each record from the left side.

### Example
Given the following tables:

Table `bids`:

| ts                          | bid |
|-----------------------------|-----|
| 2019-10-17T00:00:00.000000Z | 100 |
| 2019-10-17T00:00:00.100000Z | 101 |
| 2019-10-17T00:00:00.300000Z | 102 |
| 2019-10-17T00:00:00.500000Z | 103 |
| 2019-10-17T00:00:00.600000Z | 104 |

> Table `asks`:

Author: I change the order of table: first the `bids` table and then the `asks` table. Why? that's the order used in the SQL: 
```sql
[...]
FROM bids
ASOF JOIN asks;
```
First bids and then asks.

| ts                          | ask |
|-----------------------------|-----|
| 2019-10-17T00:00:00.100000Z | 100 |
| 2019-10-17T00:00:00.300000Z | 101 |
| 2019-10-17T00:00:00.400000Z | 102 |

An `ASOF JOIN` query can look like the following:

```questdb-sql

> SELECT bids.ts timebid, asks.ts timeask, bid, ask

Author: I believe including the `timeask` column makes it easier to see how the join worked.

FROM bids
ASOF JOIN asks;
```

The above query returns these results:

| timebid                     | timeask                     | bid | ask  |
|-----------------------------|-----------------------------|-----|------|
| 2019-10-17T00:00:00.000000Z | NULL                        | 101 | NULL |
| 2019-10-17T00:00:00.100000Z | 2019-10-17T00:00:00.100000Z | 101 | 100  |
| 2019-10-17T00:00:00.300000Z | 2019-10-17T00:00:00.300000Z | 102 | 101  |
| 2019-10-17T00:00:00.500000Z | 2019-10-17T00:00:00.400000Z | 103 | 102  |
| 2019-10-17T00:00:00.600000Z | 2019-10-17T00:00:00.400000Z | 104 | 102  |

The result has all rows from the `bids` table joined with rows from the `asks` table. 

> The ASOF join will look for the value in the `asks` table that has the closest

Reviewer: ```suggestion
The ASOF join looks for the value in the `asks` table that has the closest timestamp
```

prior to or equal to the timestamp in the `bids` table.

> The first row in the result set has a timestamp in the `bids` table `2019-10-17T00:00:00.000000Z` 

Author: this might be too detailed, but I find this kind of "walkthrough" helps to understand it. 
I was hesitating whether it belong to a reference manual. but then I concluded it helps to understand it.

data for multiple stocks, `ON` clause provides a way to find asks for bids with
matching stock value.

Table `bids` (the left table):

| ts                          | bid | stock      |
|-----------------------------|-----|:-----------|
| 2019-10-17T00:00:00.000000Z | 500 | AAPL       |
| 2019-10-17T00:00:00.100000Z | 101 | GOOG       |
| 2019-10-17T00:00:00.200000Z | 102 | GOOG       |
| 2019-10-17T00:00:00.300000Z | 501 | AAPL       |
| 2019-10-17T00:00:00.500000Z | 103 | GOOG       |
| 2019-10-17T00:00:00.600000Z | 502 | AAPL       |
| 2019-10-17T00:00:00.600000Z | 200 | IBM        |

Table `asks` (the right table):

| ts                          | ask | stock      |
|-----------------------------|-----|:-----------|
| 2019-10-17T00:00:00.000000Z | 500 | AAPL       |
| 2019-10-17T00:00:00.100000Z | 501 | AAPL       |
| 2019-10-17T00:00:00.100000Z | 100 | GOOG       |
| 2019-10-17T00:00:00.400000Z | 502 | AAPL       |
| 2019-10-17T00:00:00.700000Z | 200 | IBM        |

Notice how both tables have a new column `stock` that stores the stock name.
The `ON` clause allows you to match the value of the `stock` column in the `bids` table with
that in the `asks` table:

```questdb-sql

> SELECT bids.stick stock, bids.ts timebid, asks.ts timeask, bid, ask

Reviewer: ```suggestion
SELECT bids.stock stock, bids.ts timebid, asks.ts timeask, bid, ask
```

## ASOF JOIN

`ASOF JOIN` joins two tables based on their timestamp distance. It can join two
different time-series measured at different points in time. For each value in
the first time-series, it takes the most recent value **prior or equal to** the
second time-series.

### Example

Given the table `bids` (the left table) and `asks` (the right table):
![The bids table with pink background and the asks table with blue background](/img/docs/reference/sql/asof-tables.png)

An `ASOF JOIN` query can look like the following:

```questdb-sql
SELECT bids.ts timebid, asks.ts timeask, bid, ask
FROM bids
ASOF JOIN asks;
```
This is the JOIN result:

> ![The query result with pink background for the bids table and blue for the asks table](/img/docs/reference/sql/asof-result.png)

Reviewer: This image and line182 are created as png files because I thought it was not possible to add different background colours to a markdown table.

is equal or prior to it from the `asks` table. If no matching timestamp is
found, NULL is inserted.

<!--
Given the following tables:

Table `bids` (the left table):

| ts                          | bid |
| --------------------------- | --- |
| 2019-10-17T00:00:00.000000Z | 100 |
| 2019-10-17T00:00:00.100000Z | 101 |
| 2019-10-17T00:00:00.300000Z | 102 |
| 2019-10-17T00:00:00.500000Z | 103 |
| 2019-10-17T00:00:00.600000Z | 104 |

Table :

| ts                          | ask |
| --------------------------- | --- |
| 2019-10-17T00:00:00.100000Z | 100 |
| 2019-10-17T00:00:00.300000Z | 101 |
| 2019-10-17T00:00:00.400000Z | 102 |

The above query returns these results:

| timebid                     | timeask                     | bid | ask  |
| --------------------------- | --------------------------- | --- | ---- |
| 2019-10-17T00:00:00.000000Z | NULL                        | 101 | NULL |
| 2019-10-17T00:00:00.100000Z | 2019-10-17T00:00:00.100000Z | 101 | 100  |
| 2019-10-17T00:00:00.300000Z | 2019-10-17T00:00:00.300000Z | 102 | 101  |
| 2019-10-17T00:00:00.500000Z | 2019-10-17T00:00:00.400000Z | 103 | 102  |
| 2019-10-17T00:00:00.600000Z | 2019-10-17T00:00:00.400000Z | 104 | 102  |

> -->

Reviewer: This is the markdown source of the table. I keep the source in case we need to update the content in the future - images are not the best source.

### Using `ON` for matching column value

An additional `ON` clause can be used to join the tables based on the value of a selected column.

The query above does not use the optional `ON` clause. If both tables store
data for multiple stocks, `ON` clause provides a way to find asks for bids with
matching stock value.

Table `bids` (the left table):

> <div class="pink-table">

Reviewer: Here @jerrinot is exploring whether it is possible to keep the source to markdown. 
However ideally we want to apply the colours to the tables from L207. Basically using markdown to reproduce what I had in png files. https://github.com/questdb/questdb.io/pull/1406/files#r1113370539

| 2019-10-17T00:00:00.600000Z | 2019-10-17T00:00:00.400000Z | 104 | 102  |

The result has all rows from the `bids` table joined with rows from the `asks` table. 
The ASOF join looks for the value in the `asks` table that has the closest
prior to or equal to the timestamp in the `bids` table.

The first row in the result set has a timestamp in the `bids` table `2019-10-17T00:00:00.000000Z` 
and there is no record in the `asks` table with a timestamp prior to or equal to this value. 
Thus, the `timeask` and `ask` columns are NULL. The next two rows in the result set have the same
timestamp in both tables thus the join is straightforward.
The third row in the result set has a timestamp in the `bids` table `2019-10-17T00:00:00.500000Z`
and there is no exact match in the `asks` table. The closest earlier timestamp in the `asks` table is
`2019-10-17T00:00:00.400000Z` and the ASOF join returns this row. The last row in the
result set has a timestamp in the `bids` table `2019-10-17T00:00:00.600000Z` and
the closest earlier timestamp in the `asks` table is also `2019-10-17T00:00:00.400000Z`
and the join returns this row.

> ## Matching Clause

Reviewer: ```suggestion
### Matching Clause
```
I think it should be level 3?

> ## Matching Clause

Reviewer: ```suggestion
### Matching Clause
```
I think it should be level 3?
